### ISSUE TITLE
Investigate Grouped Time Entry feature

### ISSUE THREAD

**Issue by amulware** [toggl/mobileapp#4067]
We want to bring the Grouped Time Entry feature to our apps.

## What we have

Business logic documentation: https://www.notion.so/Time-Entries-c4bee587edc74fcb905a1844f973181e#1a130a8d88d44479a55558672a498f15

Examples: Check the Desktop and Web app, both implement this feature

## What we need

- expose this setting in the user preferences
- if enabled, merge time entries correctly in the main log
- display those groups differently
- allow the groups to be expanded/collapsed
- possibly allow the groups to be edited (not sure on this one? if so, this is a significant complication - check this with the other apps/with @maritoggl, and then get back to @amulware for the final call)
- have this UI update correctly when TEs are changed, be that triggered by the user, or sync (let's NOT have main-view crashes for half of our users this time ;) )

## How to close this issue

- [ ] determine exact requirements
- [ ] get an idea of how we'll implement it technically (check this with @heytherewill)
- [ ] analyse everything that can go wrong (we're making changes to the main log again!)
- [ ] give a rough estimate of how much work this will be, for how many people

**Comment by simonrozsival** [comment]
> possibly allow the groups to be edited

I think it would be better to avoid this in the first iteration and only add group editing if users request this feature.

> get an idea of how we'll implement it technically

I think this is very straightforward. We just group the TEs based on the criteria listed in Notion ("description, project, task, billable flag, workspace, user, tags (or lack thereof) and start date match") using Linq. The risky thing would be to change the way we update the table view source because this used to be problematic in the past when we experienced many crashes because of several bugs in this piece of code.

The way we display the grouped TEs will be different on each platform, but it should be quite similar to how we deal with the "workspaces > projects > tasks" grouping in the project/task selection views.

It might be challenging to animate the changes when two existing TEs should be merged or when a TE is moved to or from a group but it will definitely be possible to figure it out and guys with more experience with both platforms might have better estimate how hard this is going to be.

I think it would be much easier if we made the changes to both platforms at once and if we didn't have to make some temporary changes to any of the platforms because of the changes we make in the shared code. This would require having both designs ready at the same time.

I also think we don't have to wait for any designs before we start working on this feature. We could take the way the frontend grouped TEs look for now and start implementing all the changes to the underlying data structures and the UI. We can then make it look nicer when the designs are ready.

> analyse everything that can go wrong (we're making changes to the main log again!)

We used to have many crashes because of inconsistent number of sections and rows after updates on iOS (the last time the bug was fixed and all the related info can be found here: https://github.com/toggl/mobileapp/pull/2974). There were related problems on Android. We should pay close attention to the changes we make and test the app a lot when this is done.

I don't think there could be any other significant problems with this change.

**Comment by sofokli1** [comment]
This is Daneel and Foundation point of view (I leave the android to someone else)

I have split this into 2 parts:
# 1. Grouping and single TE editing

## Determine exact requirements
The issue description has plenty of info and the rest should be covered below.

## Get an idea of how we'll implement it technically

### Settings
- We need an other switch (like the manual mode switch) with tittle `Group similar time entries`. This can be done in interface builder.
- The switch value should be connect with the `SettingsViewModel` in `SettingsViewController` and stored in `IUserPreferences`.

### Log (presentation layer)
- The cell that represents the group should look the same as the single cells but with a label representing the number of timeEntries in the group. The label should have a square shape with rounded corners around it as it is in web. When opened it should have a slightly grey background and the number should have a green colour (like web).
- When the cell is tapped it should add the time entries it groups, below itself.

##### Expanded view
- The grouped timeEntries that are added to the log should look like single time entries but with a slightly grey background so that are easy to distinguish.

##### Closed view
- When the grouped timeEntry cell is pressed again the timeEntries shown below it should be removed. And the cell label indicating the number of grouped timeEntries should look as before the expansion.

##### Animation
- The animation to use should be expanding from the top when opening (adding the grouped cells) and the reverse of it when closing (removing the grouped cells) [like this](https://goo.gl/images/Q5Z8jN). Taping any of the groupChild cells should open their individual edit view. The above described design is how I see it, feedback is welcome.

### Log (Foundation)

#### - `TimeEntryViewModel`
- Add `IsExpanded` flag to keep track if the group is expanded or not
- Add `IsGroupChild` flag to distinguish from single timeEntries and from the parentGroup timeEntry. Will be used for when selecting to show edit view and setting its background colour.
- Add `IsGroupParent` flag to decide if it needs the count label in the cell. (Can also be omitted and checked by using the count of the `ChildTimeEntries` property)
- `ChildTimeEntries` list containing `TimeEntryViewModel`s that were grouped.

#### - `TimeEntriesViewModel`
- Perform the initial grouping as described in notion
- Add functions to expand/collapse a TE by modifying the corresponding `TimeEntryViewModel` flags (in case of parent) and adding/removing its child timeEntryVMs from the `TimeEntries` collection.

Changes in the `TimeEntries` collection will trigger updates to the UI accordingly, without changing the current logic for UI updates to accommodate this new feature. The idea here is to handle the grouped timeEntries similarly to how we handle single timeEntries now.

#### - `MainViewModel`
- Update `Task timeEntrySelected(long timeEntryId)` to get the whole `TimeEntryViewModel` object instead of the id only and call the the new `TimeEntriesViewModel` expand/collapse methods as needed.

### Log (Daneel)

#### - `TimeEntriesLogViewCell`
We can use the same cell for all 3 kinds of timeEntries (single, groupParent, groupChild) since the main differences is the background colour and the label showing the count of grouped timeEntries, which can be hidden.

#### - `MainViewController`
- Update `tableViewSource.ItemSelected` binding to send the whole `item` instead of just the ID.

### Analyse everything that can go wrong
Not much or nothing should go wrong with this approach.

### Rough estimate
For Daneel and Foundation changes:
1 dev 3-5 days for implementing, deciding on what should be the ID of the groupParent, bug fixes, review, fix PR, nagging people to review and re-review, possible sick days, possible unforeseen issues with the described approach.

I will leave the android part to someone else.

# 2. Group editing
I do not recommend this from the beginning since it needs some UX thinking. In case this moves forward this could be done by having a special edit view that omits info like the time but keeps all the other info like description, project, date, etc. All changes done to this view will be applied to the all the grouped time entries. This view would be triggered by either tapping the groupParent or long pressing the groupParent (depending on what is easier and more intuitive for the user).

- If we go with taping the groupParent to show the group edit view, then the expansion/collapse of the child TEs should happen when taping the label that shows the number of the child TEs (which at this point should be a button).
- If we go with the long press the groupParent to show the group edit view, then a short/normal tap should expansion/collapse of the child TEs.

Tapping on the groupChild timeEntries should behave as with the single timeEntries.

This should take around 2-3 days for 1 dev Daneel and Foundation.

@heytherewill and others feedback is welcome

Will un-assign myself so that someone else from droid can say his piece.

**Comment by simonrozsival** [comment]
Just a note on the settings: this should not be stored in `IUserPreferences`. There is a boolean property of the `/me/preferences` endpoint and so we should add a property to `IPreferences` and sync it the same way we do it for date/time formats.

**Comment by sofokli1** [comment]
👆

**Comment by maritoggl** [comment]
Confirming that there's no need for group editing during the first iteration. Keep the idea of it at the back of your head for future purposes, but this should not be a blocker. Currently, webapp has support for group edit, while desktop app doesn't.

**Comment by amulware** [comment]
If other @toggl/ios-senseis agree that this is dealt with for the foundation/iOS side, please move it to the in-review column on the iOS project.

**Comment by khoi** [comment]
comment on the proposed iOS implementation

> Add IsExpanded flag to keep track if the group is expanded or not
> ChildTimeEntries list containing TimeEntryViewModels that were grouped.

I'd like to keep the `TimeEntryViewModel` immutable and create something like `GroupedTimeEntryViewModel` which contains those properties.

And then we can update the `TimeEntriesLogViewSource` to support the new type. 

It's very similar to @sofokli1 solution, just without having to mutate the booleans

**Comment by arturs105** [comment]
I'm with @khoi - i think we shouldn't use `TimeEntryViewModel` for groups, but instead have a new view model.

**Comment by amulware** [comment]
The information above is sufficient for me, thank you.

**Comment by amulware** [comment]
One thought: Regarding whether to show all data for all time entries of an expanded group, consider that on the web you can edit time entries in place, while on mobile you can't. There is no functional need to duplicate the data in that sense.

**Comment by amulware** [comment]
Closing since everything required was covered.

**Comment by amulware** [comment]
Another thing that was not mentioned so far I believe: We may want to rxfactor the main view before we do this.
